feat(cloudfront): load external pem keys for Trusted Key Groups

robertd opened the pull request:
@njlynch Here is my first take on this... Constructive criticism is more than welcome. ✌️ 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ec86e01bb11ed32d0830cbb86ff59f60ec20520c
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6dcd06b8-4309-407c-9f89-0cd3c3e23f61%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f1053d0af4226d0f3234598386f2dd6364399e67
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=80b776ef-09c6-476e-b76d-b440666bf884%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

robertd commented:
@njlynch I've been following your guidance from https://github.com/aws/aws-cdk/pull/12743#issuecomment-771046345 . Here are the few things that I'm trying to figure out:

- I've been getting the following error: `[awslint:construct-ctor-props-optional:@aws-cdk/aws-cloudfront.PublicKey] construct "props" must be optional since all props are optional`. In this MR all the `PublicKeyProps` params are now optional, thus `awslint` suggesting that `PublicKey` needs to be optional as well in `KeyGroup`. This leads me to believe that we need to create something similar to `lambda.Code.fromInline` and `lambda.Code.fromAsset`. But in this case our "asset" would not be uploaded to S3 bucket... it would just be loaded from the filesystem, unless that S3 upload is desired feature. WDYT?
- Current implementation is working... but something that I'm not really sure about is the approach of handling `encodedKeyPath` for `fs.readFileSync` implementation. I've noticed that while writing tests I had to use relative paths and that looked weird to me. This kinda pushed me even more to think about using something similar to `lambda.Code.fromInline` and `lambda.Code.fromAsset` approach.
- PublicKey class now has bunch of `if` checks... is there a better way to approach this? I can think of few more checks (is it pem file, etc).
- What encoding `fs.readFileSync` should use when reading the .pem key in? `utf-8`?
- Do you have any design guidance we should pursue instead?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3d9c7e10b85693b291051a7153032cdd0b05c83d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9818e765-b9a0-4ae2-9eea-3d4b14e59832%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

robertd commented:
> * I've been getting the following error: `[awslint:construct-ctor-props-optional:@aws-cdk/aws-cloudfront.PublicKey] construct "props" must be optional since all props are optional`. In this MR all the `PublicKeyProps` params are now optional, thus `awslint` suggesting that `PublicKey` needs to be optional as well in `KeyGroup`. This leads me to believe that we need to create something similar to `lambda.Code.fromInline` and `lambda.Code.fromAsset`. But in this case our "asset" would not be uploaded to S3 bucket... it would just be loaded from the filesystem, unless that S3 upload is desired feature. WDYT?
> * Current implementation is working... but something that I'm not really sure about is the approach of handling `encodedKeyPath` for `fs.readFileSync` implementation. I've noticed that while writing tests I had to use relative paths and that looked weird to me. This kinda pushed me even more to think about using something similar to `lambda.Code.fromInline` and `lambda.Code.fromAsset` approach.

@njlynch These are two bullet points are kinda obsolete now after 41c1a0a. I've borrowed heavily from `lambda.Code`. Let me know what you think of overall approach.

> * PublicKey class now has bunch of `if` checks... is there a better way to approach this? I can think of few more checks (is it pem file, etc).

InlineKey class has a majority of these checks now. However, let me know what else we can check on. I'll gladly add more and update tests.

On a side note... can you please explain what's the purpose of `!Token.isUnresolved(key)` (https://github.com/aws/aws-cdk/blob/41c1a0aae275fc88e5483bc2cfcc13a63d53950d/packages/%40aws-cdk/aws-cloudfront/lib/public-key.ts#L143) and what kind of error handling we would need to put in place for a standalone check? I was thinking of separating `-----BEGIN-----` and `-----END-----` checks from Token check. Thoughts?

> * What encoding `fs.readFileSync` should use when reading the .pem key in? `utf-8`?

In addition to this... any guidelines on error handling when reading files? Should I put this in try/catch block? i.e. check if pem file is empty, check if it has `.pem` extension, etc.

> Do you have any design guidance we should pursue instead?

Let me know... If this makes sense from design perspective and if we need to go in different direction.


P.S. I think we should merge this PR first and then merge #12847. That way we can update/rebase  #12847 once this one lands on master and we won't have breaking API changes right off the bat. I'll set #12847 as draft. Thoughts?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 41c1a0aae275fc88e5483bc2cfcc13a63d53950d
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b786fc19-1c9b-48b5-9e1f-b6dcf6b8bea5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e0ee449518385b86129d8ba52a0e2191a3bff289
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=915d6130-fe12-4aa0-bcf0-773269b6bef2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8098e49f561441538a398bb4a8114ef9dee4c80a
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4b0e6b26-1613-4f91-ae9c-5118934cb7cb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

robertd commented:
Hmm... my latest builds are failing... the only thing that changed is that `1.88.0` came out in the meantime....

[Working build on `1.87.1`](https://ops-cdk-autobuild-prs-autobuildgithubco-buildlogs-scgj9jnkdtg5.s3.amazonaws.com/b786fc19-1c9b-48b5-9e1f-b6dcf6b8bea5/build.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAWYORDRBKLDYVXEZE%2F20210204%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210204T050316Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBUaCXVzLWVhc3QtMSJGMEQCIA7U3E6vKTbX%2BUYF3OZHW6eAzVXcZmS7ZAnjetFavfTcAiBGKPR8mnIv1p8nor9jtAhT98BVEl4lc3AKN0qyZljcqiqIAgj%2B%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDQ2NDgzMTg3NTE1NiIM4icTu9Pg6Oxw20uDKtwB%2BElFxyXZgJmTzMhKIgO%2F5%2BD6yUl05zhYEdqBrE2QfvjrNlJz99kB7%2FhXbdy7XwlVI6xMTMWYXp70z3er%2BwlLHf7XAJom%2BXsXdtVB%2B0zgk5SF6C3CZjmkAnDQl2%2B0oCeFYnmY%2Bg%2BZQ%2BEFDXP0YVv8aNFQuPM86RU1a04Ll5XjopL7YVIBh4jlD%2FT%2FDSqyGtdzUVBc8QncFDvxMFq4Vmmr7sGZGUiQoxh5T%2BR0G1IjH2a9vFUHNXksT3lsGzy1fyKvWL0vsf1VrlCSmHBcvHExe3csRoRtI4pkgzQVIjDM%2F%2B2ABjrhAY1vBWZH09nLxqTggpMpwplerQYvONx6pFUjRrF3C8KTNzEtIPJiGqs88%2FPYADMNSpFI5%2BSbgpizGxRMOXt7ARg8SFHs3yRXWuYHhXuU1Htn93tMDOgeFW3xfMjdBiYfKDw3f4Yyt3IF9J0syXEwdI%2Bzi97jZrmoz9m396vZzVjTbSm4EVwBZFDYFPupeLjJzMknpsIecMtMLG42bcwqx3Qv7%2FFlp15XiILJbQgsXjPpvqD7Hmhsuj0xifqwVEFBGyrBGnSA%2BnzMXWKlh%2FBJL6xgk9ok9tWRVYYwKrryjnIGww%3D%3D&X-Amz-Signature=2c453706c4390ee43f7d2ba2430d81510c42504851fc567d99a4db9c2fdc3eb0)
```
Determining baseline version...
current version: 1.87.1
current version: 1.87.1
  Current version is 1.87.1.
Using version '1.87.1' as the baseline...
Installing from NPM...
npx: installed 254 in 2.999s
npm WARN deprecated @aws-cdk/cdk-assets-schema@1.87.1: merged into @aws-cdk/cloud-assembly-schema
npm WARN deprecated @aws-cdk/aws-eks-legacy@1.87.1: Use the @aws-cdk/aws-eks module instead
npm WARN deprecated @aws-cdk/aws-dynamodb-global@1.87.1: This module has been deprecated. Use @aws-cdk/aws-dynamodb.Table with replicationRegions instead
npm WARN deprecated @aws-cdk/app-delivery@1.87.1: Use the @aws-cdk/pipelines module instead

added 184 packages, and audited 223 packages in 22s

found 0 vulnerabilities
Checking compatibility...
monocdk... OK.
@aws-cdk/alexa-ask... OK.
@aws-cdk/app-delivery... OK.
@aws-cdk/assets... OK.
@aws-cdk/aws-accessanalyzer... OK.
@aws-cdk/aws-acmpca... OK.
@aws-cdk/aws-amazonmq... OK.
@aws-cdk/aws-amplify... OK.
@aws-cdk/aws-apigateway... OK.
@aws-cdk/aws-apigatewayv2-integrations... OK.
@aws-cdk/aws-apigatewayv2... OK.
@aws-cdk/aws-appconfig... OK.
@aws-cdk/aws-appflow... OK.
@aws-cdk/aws-applicationautoscaling... OK.
@aws-cdk/aws-applicationinsights... OK.
@aws-cdk/aws-appmesh... OK.
@aws-cdk/aws-appstream... OK.
@aws-cdk/aws-appsync... OK.
@aws-cdk/aws-athena... OK.
@aws-cdk/aws-auditmanager... OK.
@aws-cdk/aws-autoscaling-common... OK.
@aws-cdk/aws-autoscaling-hooktargets... OK.
@aws-cdk/aws-autoscaling... OK.
@aws-cdk/aws-autoscalingplans... OK.
@aws-cdk/aws-backup... OK.
@aws-cdk/aws-batch... OK.
@aws-cdk/aws-budgets... OK.
@aws-cdk/aws-cassandra... OK.
@aws-cdk/aws-ce... OK.
@aws-cdk/aws-certificatemanager... OK.
@aws-cdk/aws-chatbot... OK.
@aws-cdk/aws-cloud9... OK.
@aws-cdk/aws-cloudformation... OK.
@aws-cdk/aws-cloudfront-origins... OK.
@aws-cdk/aws-cloudfront... OK.
...
```

[Non-working build on `1.88.0`](https://ops-cdk-autobuild-prs-autobuildgithubco-buildlogs-scgj9jnkdtg5.s3.amazonaws.com/4b0e6b26-1613-4f91-ae9c-5118934cb7cb/build.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAWYORDRBKLDYVXEZE%2F20210204%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210204T050206Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBUaCXVzLWVhc3QtMSJGMEQCIA7U3E6vKTbX%2BUYF3OZHW6eAzVXcZmS7ZAnjetFavfTcAiBGKPR8mnIv1p8nor9jtAhT98BVEl4lc3AKN0qyZljcqiqIAgj%2B%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDQ2NDgzMTg3NTE1NiIM4icTu9Pg6Oxw20uDKtwB%2BElFxyXZgJmTzMhKIgO%2F5%2BD6yUl05zhYEdqBrE2QfvjrNlJz99kB7%2FhXbdy7XwlVI6xMTMWYXp70z3er%2BwlLHf7XAJom%2BXsXdtVB%2B0zgk5SF6C3CZjmkAnDQl2%2B0oCeFYnmY%2Bg%2BZQ%2BEFDXP0YVv8aNFQuPM86RU1a04Ll5XjopL7YVIBh4jlD%2FT%2FDSqyGtdzUVBc8QncFDvxMFq4Vmmr7sGZGUiQoxh5T%2BR0G1IjH2a9vFUHNXksT3lsGzy1fyKvWL0vsf1VrlCSmHBcvHExe3csRoRtI4pkgzQVIjDM%2F%2B2ABjrhAY1vBWZH09nLxqTggpMpwplerQYvONx6pFUjRrF3C8KTNzEtIPJiGqs88%2FPYADMNSpFI5%2BSbgpizGxRMOXt7ARg8SFHs3yRXWuYHhXuU1Htn93tMDOgeFW3xfMjdBiYfKDw3f4Yyt3IF9J0syXEwdI%2Bzi97jZrmoz9m396vZzVjTbSm4EVwBZFDYFPupeLjJzMknpsIecMtMLG42bcwqx3Qv7%2FFlp15XiILJbQgsXjPpvqD7Hmhsuj0xifqwVEFBGyrBGnSA%2BnzMXWKlh%2FBJL6xgk9ok9tWRVYYwKrryjnIGww%3D%3D&X-Amz-Signature=77e62ef618ead4e2146c3b28fa06ba0de2d5015903158b83a1b7daf584397066)
```
Determining baseline version...
current version: 1.88.0
current version: 1.88.0
  Current version is 1.88.0.
Using version '1.88.0' as the baseline...
Installing from NPM...
npx: installed 254 in 2.952s
npm WARN deprecated @aws-cdk/cdk-assets-schema@1.88.0: merged into @aws-cdk/cloud-assembly-schema
npm WARN deprecated @aws-cdk/aws-dynamodb-global@1.88.0: This module has been deprecated. Use @aws-cdk/aws-dynamodb.Table with replicationRegions instead
npm WARN deprecated @aws-cdk/aws-eks-legacy@1.88.0: Use the @aws-cdk/aws-eks module instead
npm WARN deprecated @aws-cdk/app-delivery@1.88.0: Use the @aws-cdk/pipelines module instead

added 184 packages, and audited 223 packages in 18s

found 0 vulnerabilities
Checking compatibility...
monocdk... OK.
@aws-cdk/alexa-ask... OK.
@aws-cdk/app-delivery... OK.
@aws-cdk/assets... OK.
@aws-cdk/aws-accessanalyzer... OK.
@aws-cdk/aws-acmpca... OK.
@aws-cdk/aws-amazonmq... OK.
@aws-cdk/aws-amplify... OK.
@aws-cdk/aws-apigateway... OK.
@aws-cdk/aws-apigatewayv2-integrations... OK.
@aws-cdk/aws-apigatewayv2... OK.
@aws-cdk/aws-appconfig... OK.
@aws-cdk/aws-appflow... OK.
@aws-cdk/aws-applicationautoscaling... OK.
@aws-cdk/aws-applicationinsights... OK.
@aws-cdk/aws-appmesh... OK.
@aws-cdk/aws-appstream... OK.
@aws-cdk/aws-appsync... OK.
@aws-cdk/aws-athena... OK.
@aws-cdk/aws-auditmanager... OK.
@aws-cdk/aws-autoscaling-common... OK.
@aws-cdk/aws-autoscaling-hooktargets... OK.
@aws-cdk/aws-autoscaling... OK.
@aws-cdk/aws-autoscalingplans... OK.
@aws-cdk/aws-backup... OK.
@aws-cdk/aws-batch... OK.
@aws-cdk/aws-budgets... OK.
@aws-cdk/aws-cassandra... OK.
@aws-cdk/aws-ce... OK.
@aws-cdk/aws-certificatemanager... OK.
@aws-cdk/aws-chatbot... OK.
@aws-cdk/aws-cloud9... OK.
@aws-cdk/aws-cloudformation... OK.
@aws-cdk/aws-cloudfront-origins... OK.
@aws-cdk/aws-cloudfront... CHANGES.
Original assembly: @aws-cdk/aws-cloudfront@1.88.0
Updated assembly:  @aws-cdk/aws-cloudfront@0.0.0
API elements with incompatible changes:
err  - IFACE @aws-cdk/aws-cloudfront.PublicKeyProps: property encodedKey, string is not assignable to @aws-cdk/aws-cloudfront.Key: input to @aws-cdk/aws-cloudfront.PublicKey.<initializer> [strengthened:@aws-cdk/aws-cloudfront.PublicKeyProps]
...
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a500b5d9237b97a8005b54a0485233e1c5151f13
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=92df1c5f-6426-4706-8921-0a2ee5f377ab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

robertd commented:
@njlynch Do you happen to know why `package.json` is full of `0.0.0`s? Is it possible that’s why builds started failing in with the release of `1.88.0`?

njlynch commented:
> Still trying to figure out why builds are failing.
> `err  - IFACE @aws-cdk/aws-cloudfront.PublicKeyProps: property encodedKey, string is not assignable to @aws-cdk/aws-cloudfront.Key: input to @aws-cdk/aws-cloudfront.PublicKey.<initializer> [strengthened:@aws-cdk/aws-cloudfront.PublicKeyProps]`

The build is failing because we can't make backwards-incompatible changes to (non-experimental) classes (in non-experimental modules). As the PublicKey L2 was released, it now is a contract that we can't break subsequently by changing the type of one of the properties.

If we want to add this functionality now, the options are:
1. Add the new property for file name/path, make both fields optional, suppress the linter warnings, and add checks in the constructor that confirm that exactly one of the key fields has been provided. I think this is similar to your first approach here (and what I mentioned in #12743). The downside, of course, to this approach is that the compiler isn't forcing that one of these fields is present, so we're pushing validation to 'runtime' checks rather than having the compiler help us do the right thing.
2. A mix of the above and what you've done here. We can create a new field with the new type, deprecate the existing string-based one, and make both optional. We still suffer from 'runtime' checks, but can remove the old field in our v2 launch.
3. Do nothing for a bit. This construct was _just_ released. Wait for some feedback from the community on what makes sense. Maybe folks actually want to feed in a public key from SSM, or Secrets Manager, or something else, rather than local files. Maybe some helper construct that generates both private & public key, provides the public key to CloudFront and stores the private in Secrets Manager. Once we have a slightly better picture of what users need the most, we can attack this with more data.

> On a side note... can you please explain what's the purpose of !Token.isUnresolved(key)

Sure. https://docs.aws.amazon.com/cdk/latest/guide/tokens.html is a handy primer. Short version -- any string (or number, or list) in the CDK _might_ actually be a Token which references some other value from some other construct/resource. So rather than the string being `----BEGIN PUBLIC KEY----....`, it might be `${TOKEN[SecretsManager.Value.1234]}`, making "normal" string comparison/manipulation useless. In those cases, we usually opt to trust the input and skip whatever verification check we're performing.

robertd commented:
@njlynch Thanks again for explaining all this. I figured it was probably a breaking contract. Makes total sense. 
 
For the future reference, lesson learned on my part, maybe I should’ve created these two L2 constructs as experimental first and let the dust settle a bit... before changing the Implementation :)

In our use case (multiple teams using same AWS acct) we wanted to create bunch of trusted key groups (and pub keys) using factory pattern from file system and that guided my design of this PR. But I like the ideas you mentioned in option 3 so let’s wait for it then. 

You can go ahead and review/merge #12847 so we can at least use new constructs in Distribution. 

I’ll leave this PR as draft and come back to it after we get some feedback from the community.

njlynch commented:
@robertd  - Would you be okay just closing out this PR until we've received some feedback and ready to implement something like this? I keep seeing this in my "queue" and thinking I need to review it. :)  We can always re-open it later.

robertd commented:
@njlynch 👍
